components/emoji_picker/emoji_picker.jsx

```diff
@@ -149,6 +149,61 @@ export default class EmojiPicker extends React.PureComponent {
         customEmojisEnabled: false,
     };
 
+    static getEmojis(props, state) {
+        const {categories, allEmojis} = state;
+        const emojiMap = props.emojiMap;
+        const customEmojiMap = emojiMap.customEmojis;
+
+        for (const category of Object.keys(categories)) {
+            let categoryEmojis = [];
+            if (category === 'recent' && props.recentEmojis.length) {
+                const recentEmojis = [...props.recentEmojis].reverse();
+                categoryEmojis = recentEmojis.filter((name) => {
+                    return emojiMap.has(name);
+                }).map((name) => {
+                    return emojiMap.get(name);
+                });
+            } else {
+                const indices = Emoji.EmojiIndicesByCategory.get(category) || [];
+                categoryEmojis = indices.map((index) => Emoji.Emojis[index]);
+                if (category === 'custom') {
+                    categoryEmojis = categoryEmojis.concat([...customEmojiMap.values()]);
+                }
+            }
+            categories[category].emojiIds = categoryEmojis.map((emoji) => getEmojiFilename(emoji));
+            for (let i = 0; i < categoryEmojis.length; i++) {
+                const currentEmoji = categoryEmojis[i];
+                const fileName = getEmojiFilename(currentEmoji);
+                allEmojis[fileName] = {
+                    ...currentEmoji,
+                    visible: false,
+                    offset: null,
+                };
+                if (!currentEmoji.filename) {
+                    // if custom emoji, set proper attributes
+                    allEmojis[fileName] = {
+                        ...allEmojis[fileName],
+                        aliases: [currentEmoji.name],
+                        category: 'custom',
+                        filename: fileName,
+                    };
+                }
+            }
+        }
+
+        return {categories, allEmojis};
+    }
+
+    static getDerivedStateFromProps(props, state) {
+        let updatedState = {emojiMap: props.emojiMap};
+        if (JSON.stringify(Object.keys(state.categories)) !== state.categoryKeys || props.emojiMap !== state.emojiMap) {
```

Comment:
I don't like the way that this looks since `getState` does a lot of weird things like modifying an object in the state without necessarily returning a new reference to the object, but that's how it worked before, so I don't think we need to clean that up with this PR. The changes made by this PR seem like they should work based on some brief testing. No further changes are required here, but I want to make note of this area for when we next clean up this component.

Reply:
Also for QA, when you review this, can you spend a bit longer on the emoji picker to make sure that it still functions correctly? The way the original code was written here makes me a bit nervous about it, but everything seems to work the same here as it did before.